### PR TITLE
Adding Metadata to search results.

### PR DESCRIPTION
**Pull request by mhilton**
Search results can now have metadata attached to them using the same mechanism
as the /meta handler uses. The change also includes:
- Some cleanup of the elasticsearch interface
- Search Time and Total in search response
- More tests

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/414/

**Comment by mhilton** [comment]
This one's a bit big, sorry. To QA (Assuming you already have a full mongoDB):
1. Delete any existing index `http DELETE :9200/charmstore`
2. Populate the search database `essync -settings=internal/elasticsearch/definiitions/index.json` -mapping=internal/elasticsearch/definiitions/entity.json charmd.yaml`
3. Run charmstore`charmd charmd.yaml`
4. Search to your hearts content`http :8080/v4/search?name=mediawiki&include=archive-size`

File: cmd/essync/main.go [hunk 1]
```diff
@@ -63,10 +66,60 @@ func populate(confPath string) error {
 	}
 	defer session.Close()
 	db := session.DB("juju")
-	store, err := charmstore.NewStore(db, &charmstore.StoreElasticSearch{es, *index})
+	store, err := charmstore.NewStore(db, &charmstore.StoreElasticSearch{es.Index(*index)})
 	if err != nil {
 		return errgo.Notef(err, "unable to create store for ESSync")
 	}
+	if *settings != "" {
+		err = writeSettings(es, *index, *settings)
```
Comment (rogpeppe) [inline cmd/essync/main.go]: minor point:

   if err := writeSettings(...); err != nil {
   }

is preferable when reasonable. saves looking for potential side effects.

File: cmd/essync/main.go [hunk 2]
```diff
@@ -63,10 +66,60 @@ func populate(confPath string) error {
...
+		}
+	}
 	logger.Debugf("starting export to Elastic Search")
 	return store.ExportToElasticSearch()
 }
+
+func writeSettings(es *elasticsearch.Database, index, fn string) error {
+	f, err := os.Open(fn)
+	if err != nil {
+		return errgo.NoteMask(err, "cannot read index settings")
```
Comment (rogpeppe) [inline cmd/essync/main.go]: Or Notef.
I tend only to use NoteMask when I want to specifically allow an error type.

File: cmd/essync/main.go [hunk 3]
```diff
@@ -63,10 +66,60 @@ func populate(confPath string) error {
...
+	}
+	err = es.PutIndex(index, data)
+	if err != nil {
+		return errgo.NoteMask(err, "cannot set index settings")
+	}
+	return nil
+}
+
+func writeMapping(es *elasticsearch.Database, index, type_, fn string) error {
```
Comment (rogpeppe) [inline cmd/essync/main.go]: s/type_/indexType/
s/fn/path/ (or fileName)
?
perhaps a doc comment might be useful here.

File: internal/charmstore/search.go [hunk 4]
```diff
@@ -18,8 +21,7 @@ import (
 // elasticsearch is not configured, allowing them to be safely called even if
 // it is not enabled in this service.
 type StoreElasticSearch struct {
-	*elasticsearch.Database
-	Index string
+	*elasticsearch.Index
```
Comment (rogpeppe) [inline internal/charmstore/search.go:24]: Nice change, thanks.

File: internal/charmstore/search.go [hunk 5]
```diff
@@ -27,15 +29,20 @@ const typeName = "entity"
...
+	return ses.PutDocument(typeName, ses.getID(entity), entity)
 }
 
-// getID returns the id to be used for indexing the entity into ElasticSearch.
+// getID returns an ID for the elasticsearch document based on the contents of the
+// mongoDB document. This is to allow elasticsearch documents to be replaced with
+// updated versions when charm data is changed.
 func (ses *StoreElasticSearch) getID(entity *mongodoc.Entity) string {
-	return url.QueryEscape(entity.URL.String())
+	b := sha1.Sum([]byte(entity.URL.String()))
```
Comment (rogpeppe) [inline internal/charmstore/search.go:42]: Any particular reason to use a hash here rather than, say charm.Quote(entity.URL.String()) ?

File: internal/charmstore/search.go [hunk 6]
```diff
@@ -43,20 +50,29 @@ func (ses *StoreElasticSearch) getID(entity *mongodoc.Entity) string {
 // http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/query-dsl.html
 // It returns a slice containing each Id of the matching documents returned by
```
Comment (rogpeppe) [inline internal/charmstore/search.go]: This doc comment appears to be out of date now.

File: cmd/essync/main.go (hunk 1 above)
Comment (mhilton) [inline cmd/essync/main.go]: Sorry you mentioned that last time and I forgot to change it.

File: internal/charmstore/search.go (hunk 5 above)
Comment (mhilton) [inline internal/charmstore/search.go:42]: I wanted to remove the temptation to use it as a meaningful value when processing the results. That ID is for elasticsearch to tell the difference between things not for us to get information from, so by making it opaque now it can potentially be changed in future without breaking code that uses it incorrectly.

File: internal/elasticsearch/elasticsearch.go [hunk 7]
```diff
@@ -25,125 +27,210 @@ import (
...
+
 // EnsureID tests to see a document of the given index, type_, and id exists
 // in ElasticSearch.
 func (db *Database) EnsureID(index, type_, id string) (bool, error) {
-	// TODO (bac) We should limit the fields to the id to avoid retrieving
-	// data we don't use.
-	response, err := http.Get(db.url(index, type_, id, "_source"))
-	if err != nil {
-		return false, errgo.Mask(err)
+	if err := db.get(db.url(index, type_, id)+"?_source=false", nil, nil); err != nil {
```
Comment (urosj) [inline internal/elasticsearch/elasticsearch.go:57]: This should also be used in the /debug/status call when ES is not nil. Will add a todo.

**Comment by urosj** [comment]
:+1: 

Huge branch! I did QA - but due to lack of time I just skimmed through code. Looks fine on the first glance.

File: internal/charmstore/search.go [hunk 8]
```diff
@@ -169,8 +192,8 @@ func createSearchDSL(sp SearchParams) string {
 // for details of how filters are specified in the API. For each key in f a filter is
 // created that matches any one of the set of values specified for that key.
 // The created filter will only match when at least one of the requested values
-// matches for all of the requested keys. createFilters will silently skip over any
-// filter names that are not mapped.
+// matches for all of the requested keys. Any filter names that are not defined
+// in the filters map will be silenty skipped.
```
Comment (rogpeppe) [inline internal/charmstore/search.go]: silently

File: internal/charmstore/search.go [hunk 9]
```diff
@@ -214,12 +237,29 @@ func descriptionFilter(value string) elasticsearch.Filter {
 	}
 }
 
+// nameFilter generates a filter that will match against the
+// name of the charm or bundle.
+func nameFilter(value string) elasticsearch.Filter {
+	// expand any * wildcards in the name filter
```
Comment (rogpeppe) [inline internal/charmstore/search.go]: Is this behaviour (the \* wildcards) documented somewhere?

Comment (rogpeppe) [inline internal/charmstore/search.go]: Also, shouldn't you quota meta characters in the value other than \* ?

File: internal/charmstore/search.go [hunk 10]
```diff
@@ -248,26 +288,36 @@ func summaryFilter(value string) elasticsearch.Filter {
...
-		elasticsearch.TermFilter{
-			Field: "CharmMeta.Categories",
-			Value: value,
-		},
-		elasticsearch.TermFilter{
-			Field: "BundleData.Tags",
-			Value: value,
-		},
+	tags := strings.Split(value, " ")
+	af := make(elasticsearch.AndFilter, 0, len(tags))
```
Comment (rogpeppe) [inline internal/charmstore/search.go:289]: will this do the right thing when value is "" ?

File: internal/elasticsearch/elasticsearch.go [hunk 11]
```diff
@@ -25,125 +27,210 @@ import (
...
+func (db *Database) GetDocument(index, type_, id string, v interface{}) error {
+	if err := db.get(db.url(index, type_, id, "_source"), nil, v); err != nil {
 		return errgo.Mask(err)
 	}
-	if status != http.StatusOK {
-		return errgo.Newf("ElasticSearch GET response status: %d, body: %s", status, body)
+	return nil
+}
+
+// ListAllIndexes does a GET request to the Database and parses the string
```
Comment (rogpeppe) [inline internal/elasticsearch/elasticsearch.go]: Most of this is implementation detail which doesn't deserve to be in a doc comment.

```
// ListAllIndexes returns the names of all the indexes in the database.
// http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/_list_all_indexes.html
```

should be sufficient, I think.

File: internal/elasticsearch/elasticsearch.go [hunk 12]
```diff
@@ -25,125 +27,210 @@ import (
...
+// http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/_list_all_indexes.html
+func (db *Database) ListAllIndexes() ([]string, error) {
+	var result map[string]interface{}
+	if err := db.get(db.url("_aliases"), nil, &result); err != nil {
+		return nil, errgo.Mask(err)
 	}
-	err = json.Unmarshal(body, doc)
-	if err != nil {
-		return errgo.Notef(err, "cannot unmarshal body: %s", body)
+	var indices []string
```
Comment (rogpeppe) [inline internal/elasticsearch/elasticsearch.go]: s/indices/indexes/

might as well be consistent in the spelling.

File: internal/elasticsearch/elasticsearch.go [hunk 13]
```diff
@@ -25,125 +27,210 @@ import (
...
 	}
-	if status != http.StatusCreated {
-		return "", errgo.Newf("ElasticSearch POST response status: %d, body: %s", status, body)
-	}
-	var respdata map[string]interface{}
-	if err := json.Unmarshal(body, &respdata); err != nil {
-		return "", errgo.Notef(err, "cannot unmarshal body")
-	}
-	return respdata["_id"].(string), nil
+	return resp["_id"].(string), nil
```
Comment (rogpeppe) [inline internal/elasticsearch/elasticsearch.go]: This will panic if the response isn't the expected string.
Better to return an error in this case, I think.

File: internal/elasticsearch/elasticsearch.go [hunk 14]
```diff
@@ -25,125 +27,210 @@ import (
...
+// into tha value it points to.
+func (db *Database) delete(url string, body, v interface{}) error {
+	return db.do("DELETE", url, body, v)
+}
+
+// get makes a GET request to the database url. A non-nil body will be
+// sent with the request and if v is not nill then the response will be unmarshaled
+// into tha value it points to.
+func (db *Database) get(url string, body, v interface{}) error {
+	return db.do("get", url, body, v)
```
Comment (rogpeppe) [inline internal/elasticsearch/elasticsearch.go]: Shouldn't this be "GET" ?

File: params/params.go [hunk 15]
```diff
@@ -104,5 +104,12 @@ type SearchResult struct {
 	Meta map[string]interface{} `json:",omitempty"`
 }
 
+// SearchResponse holds the response from a search operation.
+type SearchResponse struct {
+	SearchTime time.Duration
```
Comment (rogpeppe) [inline params/params.go:115]: This might be hard to parse in Javascript. Do we care though?

**Comment by rogpeppe** [comment]
LGTM modulo some remarks above.
Thanks a lot, this is really nice.

File: cmd/essync/main.go (hunk 1 above)
Comment (mhilton) [inline cmd/essync/main.go]: done

File: cmd/essync/main.go (hunk 2 above)
Comment (mhilton) [inline cmd/essync/main.go]: done

File: cmd/essync/main.go (hunk 3 above)
Comment (mhilton) [inline cmd/essync/main.go]: done, although I named the variable documentType rather than indexType as that describes it better.

File: internal/charmstore/search.go (hunk 6 above)
Comment (mhilton) [inline internal/charmstore/search.go]: done

File: internal/charmstore/search.go (hunk 8 above)
Comment (mhilton) [inline internal/charmstore/search.go]: well spotted thanks.

File: internal/charmstore/search.go (hunk 9 above)
Comment (mhilton) [inline internal/charmstore/search.go]: The documentation is here https://docs.google.com/a/canonical.com/document/d/1eo44uCVJfha6r0v2Jz-QxbbkHtsXrabTpj_Q_ulcis4/edit#heading=h.8jimvw5a50mr

Comment (mhilton) [inline internal/charmstore/search.go]: done

File: internal/charmstore/search.go (hunk 10 above)
Comment (mhilton) [inline internal/charmstore/search.go:289]: I don;t think it did, I have made it skip empty strings.

File: internal/elasticsearch/elasticsearch.go (hunk 11 above)
Comment (mhilton) [inline internal/elasticsearch/elasticsearch.go]: done.

File: internal/elasticsearch/elasticsearch.go (hunk 13 above)
Comment (mhilton) [inline internal/elasticsearch/elasticsearch.go]: I've replaced the map with a structure to unmarshal into.

File: internal/elasticsearch/elasticsearch.go (hunk 14 above)
Comment (mhilton) [inline internal/elasticsearch/elasticsearch.go]: It worked, but I'm not sure it should. Changed.

File: params/params.go (hunk 15 above)
Comment (mhilton) [inline params/params.go:115]: It gets returned as a count in nanoseconds, which is fairly easy to parse, so I'm not going to worry.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/417/

File: internal/elasticsearch/elasticsearch.go [hunk 16]
```diff
@@ -154,144 +243,104 @@ func (db *Database) url(pathParts ...string) string {
...
+	}
+	return s
 }
 
-type ErrNotFound struct {
-	errgo.Err
+// EscapeRegexp returns the supplied string with any special characters escaped.
+// A regular expression match on the returned string will match exactly the characters
+// in the supplied string.
+func EscapeRegexp(s string) string {
```
Comment (rogpeppe) [inline internal/elasticsearch/elasticsearch.go:318]: just use regexp.QuoteMeta(s)
?

File: internal/elasticsearch/elasticsearch_test.go [hunk 17]
```diff
@@ -7,9 +7,10 @@ import (
 	"testing"
 
 	jujutesting "github.com/juju/testing"
+	//	jc "github.com/juju/testing/checkers"
```
Comment (rogpeppe) [inline internal/elasticsearch/elasticsearch_test.go]: d

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/419/

File: internal/elasticsearch/elasticsearch.go (hunk 16 above)
Comment (mhilton) [inline internal/elasticsearch/elasticsearch.go:318]: elasticsearch regular expressions use a different set of metacharacters to the go ones, so that wouldn't work.

File: internal/charmstore/search_test.go [hunk 18]
```diff
@@ -91,315 +75,266 @@ func (s *StoreSearchSuite) addCharmsToStore(store *Store) {
...
-		},
-	}
-	ids, err := s.store.Search(sp)
-	c.Assert(err, gc.IsNil)
-	c.Assert(ids, jc.SameContents, []string{
-		url.QueryEscape(exportTestCharms["wordpress"]),
-		url.QueryEscape(exportTestCharms["mysql"]),
-		url.QueryEscape(exportTestCharms["varnish"]),
-	})
+var ResultsContain gc.Checker = &resultsContainChecker{
```
Comment (rogpeppe) [inline internal/charmstore/search_test.go]: Could we have a doc comment here please, describing the
types of the arguments to the checker?

File: internal/elasticsearch/elasticsearch_test.go (hunk 17 above)
Comment (mhilton) [inline internal/elasticsearch/elasticsearch_test.go]: done

File: internal/charmstore/search_test.go (hunk 18 above)
Comment (rogpeppe) [inline internal/charmstore/search_test.go]: tbh though, it's awkward enough writing a Checker that I'd probably
avoid doing it except when it's going to be useful in lots of places.

  func assertResultsContain(*gc.C, SearchResult, []string)

is probably nicer overall.

One other thing: the "conjtain" in ResultsContain seems to imply that the results
can be a superset of the expected values. By my superficial
reading of the code, that isn't actually the case... or is it?

File: internal/elasticsearch/elasticsearch.go [hunk 19]
```diff
@@ -154,144 +243,104 @@ func (db *Database) url(pathParts ...string) string {
...
 }
 
-// Delete makes a DELETE request to the database using the path parameter.
-func (db *Database) delete(path string) (*http.Response, error) {
-	req, err := http.NewRequest("DELETE", db.url(path), nil)
-	if err != nil {
-		return nil, err
+// Index creates a reference to an index in the elasticsearch database.
+func (db *Database) Index(name string) *Index {
+	if db == nil {
```
Comment (rogpeppe) [inline internal/elasticsearch/elasticsearch.go]: Why are we doing this test here but not in any of the other Database methods?

File: internal/v4/search.go [hunk 20]
```diff
@@ -22,17 +22,26 @@ func (h *Handler) serveSearch(w http.ResponseWriter, req *http.Request) (interfa
...
-		sr := params.SearchResult{Id: id}
-		// TODO (mhilton) get metadata
-		results = append(results, sr)
+	response := params.SearchResponse{
+		SearchTime: results.SearchTime,
+		Total:      results.Total,
+		Results:    make([]params.SearchResult, 0, len(results.Results)),
 	}
-	return results, nil
+	for _, ref := range results.Results {
```
Comment (rogpeppe) [inline internal/v4/search.go:35]:
```
// TODO do this concurrently.
```

?

File: params/params.go (hunk 15 above)
Comment (rogpeppe) [inline params/params.go:115]: Ah, cool. I thought perhaps it marshaled as its string value.

**Comment by rogpeppe** [comment]
LGTM modulo a few more comments above. Thanks!

File: internal/elasticsearch/elasticsearch.go (hunk 19 above)
Comment (mhilton) [inline internal/elasticsearch/elasticsearch.go]: That's because the Index method is to create a reference to access the database with a canned index, which is the only way we ever do it. This check means that if the database isn't configured the index won't be either. For all other methods it is a failure.

File: internal/v4/search.go (hunk 20 above)
Comment (mhilton) [inline internal/v4/search.go:35]: done

File: internal/charmstore/search_test.go (hunk 18 above)
Comment (mhilton) [inline internal/charmstore/search_test.go]: done

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/422/

File: internal/elasticsearch/elasticsearch.go (hunk 19 above)
Comment (rogpeppe) [inline internal/elasticsearch/elasticsearch.go]: It still seems a bit odd to me that we're calling methods on a nil Database, but
at the least this behaviour should be documented.

**Comment by rogpeppe** [comment]
Still LGTM with some a doc comment on the Index method. FWIW can't we just put an "if" statement around the call to Index?

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/423/

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/424/

**Comment by mhilton** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge
